### PR TITLE
Disable heap contraction for implicit aggressive GC

### PR DESCRIPTION
**Pull request by dmitripivkine**
There is no reason to contract heap for implicit aggressive GC, it is going to be expanded most likely.

Issue https://github.com/eclipse-openj9/openj9/issues/17574

### CODE REVIEW

**Comment by dmitripivkine** [comment]
@amicic

**Review by amicic — APPROVED** [review]

File: gc/base/GCCode.cpp [hunk 1]
```diff
@@ -184,6 +184,46 @@ MM_GCCode::isAggressiveGC() const
...
+	case J9MMCONSTANT_IMPLICIT_GC_EXCESSIVE:
+	case J9MMCONSTANT_IMPLICIT_GC_PERCOLATE_AGGRESSIVE:
+		implicitAggressiveGC = true;
+		break;
+	case J9MMCONSTANT_EXPLICIT_GC_NATIVE_OUT_OF_MEMORY:
+	case J9MMCONSTANT_EXPLICIT_GC_RASDUMP_COMPACT:
+	case J9MMCONSTANT_EXPLICIT_GC_SYSTEM_GC:
+#if defined(OMR_GC_IDLE_HEAP_MANAGER)
+	case J9MMCONSTANT_EXPLICIT_GC_IDLE_GC:
+#endif
```
Comment (babsingh) [inline gc/base/GCCode.cpp:207]:
```suggestion
#endif /* defined(OMR_GC_IDLE_HEAP_MANAGER) */
```

File: gc/base/GCCode.cpp [hunk 2]
```diff
@@ -184,6 +184,46 @@ MM_GCCode::isAggressiveGC() const
...
+	case J9MMCONSTANT_EXPLICIT_GC_PREPARE_FOR_CHECKPOINT:
+	case J9MMCONSTANT_EXPLICIT_GC_NOT_AGGRESSIVE:
+	case J9MMCONSTANT_IMPLICIT_GC_DEFAULT:
+	case J9MMCONSTANT_IMPLICIT_GC_PERCOLATE:
+	case J9MMCONSTANT_IMPLICIT_GC_PERCOLATE_CRITICAL_REGIONS:
+	case J9MMCONSTANT_IMPLICIT_GC_PERCOLATE_UNLOADING_CLASSES:
+	case J9MMCONSTANT_IMPLICIT_GC_COMPLETE_CONCURRENT:
+#if defined(OMR_GC_CONCURRENT_SCAVENGER)
+	case J9MMCONSTANT_IMPLICIT_GC_PERCOLATE_ABORTED_SCAVENGE:
+#endif
```
Comment (babsingh) [inline gc/base/GCCode.cpp:217]:
```suggestion
#endif /* defined(OMR_GC_CONCURRENT_SCAVENGER) */
```

File: gc/base/GCCode.cpp [hunk 3]
```diff
@@ -184,6 +184,46 @@ MM_GCCode::isAggressiveGC() const
 	return aggressiveGC;
 }
 
+/**
+ * Determine if the GC is implicit aggressive.
+ * @return true if the gc code indicates an implicit aggressive GC
```
Comment (babsingh) [inline gc/base/GCCode.cpp:189]:
```suggestion
 * @return true if the GC code indicates an implicit aggressive GC
```

File: gc/base/GCCode.hpp [hunk 4]
```diff
@@ -62,6 +62,12 @@ class MM_GCCode {
 	 */
 	bool isAggressiveGC() const;
 	
+	/**
+	 * Determine if the GC is implicit aggressive.
+	 * @return true if the gc code indicates an implicit aggressive GC
```
Comment (babsingh) [inline gc/base/GCCode.hpp:67]:
```suggestion
	 * @return true if the GC code indicates an implicit aggressive GC
```

File: gc/base/j9mm.tdf [hunk 5]
```diff
@@ -1016,3 +1016,6 @@ TraceEvent=Trc_MM_ParallelDispatcher_contractThreadPool_Success noEnv Overhead=1
 TraceExit=Trc_MM_ParallelDispatcher_contractThreadPool_Exit noEnv Overhead=1 Level=1 Group=dispatcher Template="contractThreadPool Exit: gcThreadCount: %zu"
 
 TraceException=Trc_MM_ParallelDispatcher_internalStartupThreads_Failed noEnv Overhead=1 Level=1 Group=dispatcher Template="Failed to startup threads: workerThreadCount: %zu, maxWorkerThreadIndex: %zu, _threadShutdownCount: %zu"
+
+TraceExit=Trc_MM_MemorySubSpaceUniSpace_timeForHeapContract_Exit8 Overhead=1 Level=1 Group=resize Template="MM_MemorySubSpace_timeForHeapContract Exit8 Heap cannot contract in implicit aggressive GC"
+TraceExit=Trc_MM_MemorySubSpaceUniSpace_timeForHeapContract_Exit9 Overhead=1 Level=1 Group=resize Template="MM_MemorySubSpace_timeForHeapContract Exit9 Contraction required due SoftMX request, size = %zu bytes"
```
Comment (babsingh) [inline gc/base/j9mm.tdf:1021]:
```suggestion
TraceExit=Trc_MM_MemorySubSpaceUniSpace_timeForHeapContract_Exit9 Overhead=1 Level=1 Group=resize Template="MM_MemorySubSpace_timeForHeapContract Exit9 Contraction required due to SoftMX request, size = %zu bytes"
```

File: gc/base/MemorySubSpaceUniSpace.cpp [hunk 6]
```diff
@@ -257,10 +257,16 @@ MM_MemorySubSpaceUniSpace::timeForHeapContract(MM_EnvironmentBase *env, MM_Alloc
...
 
-	/* Don't shrink if we have not met the allocation request
-	 * ..we will be expanding soon if possible anyway
+	/* Don't shrink if this is implicit aggressive GC or we have not met the allocation request,
+	 * we will be expanding soon if possible anyway
 	 */
 	if (allocDescription) {
+		if (env->_cycleState->_gcCode.isImplicitAggressiveGC()) {
+			Trc_MM_MemorySubSpaceUniSpace_timeForHeapContract_Exit8(env->getLanguageVMThread());
+			_contractionSize = 0;
```
Comment (babsingh) [inline gc/base/MemorySubSpaceUniSpace.cpp:266]: The exit tracepoint should be last operation before the return statement. Is there a reason for setting `_contractionSize` after the exit tracepoint? This is also done at other places in this function.
```suggestion
			_contractionSize = 0;
			Trc_MM_MemorySubSpaceUniSpace_timeForHeapContract_Exit8(env->getLanguageVMThread());
```

**Comment by dmitripivkine** [comment]
I think all comments are addressed.

File: gc/base/MemorySubSpaceUniSpace.cpp (hunk 6 above)
Comment (babsingh) [reply gc/base/MemorySubSpaceUniSpace.cpp]: We missed one on lines 343-344.

**Comment by babsingh** [comment]
jenkins build all

**Comment by babsingh** [comment]
Line endings check failed:
```
Found the following whitespace problems:
###################################
gc/base/MemorySubSpaceUniSpace.cpp:344: space before tab in indent.
+	 		Trc_MM_MemorySubSpaceUniSpace_timeForHeapContract_Exit6(env->getLanguageVMThread(), freeBytesAtSystemGCStart, minimumFree);
###################################
```

**Comment by babsingh** [comment]
jenkins build all

**Review by babsingh — APPROVED** [review]

**Pull request merged**